docs/proposals/20210716-async-azure-resource-creation-deletion.md

```diff
@@ -0,0 +1,224 @@
+
+```go
+type Futures []Future
+
+// AzureClusterStatus defines the observed state of AzureCluster.
+type AzureClusterStatus struct {
+    /*
+      Other fields omitted for brevity    
+    */
+
+    // LongRunningOperationStates saves the states for Azure long-running operations so they can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationStates Futures `json:"longRunningOperationState,omitempty"`
+}
+
+// AzureMachineStatus defines the observed state of AzureMachine.
+type AzureMachineStatus struct {
+    /*
+      Other fields omitted for brevity    
+    */
+
+    // LongRunningOperationStates saves the states for Azure long-running operations so they can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationStates Futures `json:"longRunningOperationState,omitempty"`
+}
+
+// AzureMachinePool defines the observed state of AzureMachinePool.
+type AzureMachinePool struct {
+    /*
+      Other fields omitted for brevity    
+    */
+
+    // LongRunningOperationStates saves the states for Azure long-running operations so they can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationStates Futures `json:"longRunningOperationState,omitempty"`
+}
+
+// AzureMachinePoolMachineStatus defines the observed state of AzureMachinePoolMachine.
+type AzureMachinePoolMachine struct {
+    /*
+      Other fields omitted for brevity
+    */
+
+    // LongRunningOperationStates saves the states for Azure long-running operations so they can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationStates Futures `json:"longRunningOperationState,omitempty"`
+}
+
+```
+
+#### <a name='ProposedControllerChanges'></a>Proposed Controller Changes
+
+* Add an `AzureClientTimeout` which is the duration after which an Azure operation is considered a long running operation which should be handled asynchronously. Proposed starting value is 5 seconds.
+* For each Azure API call which returns a Future, wait for the operation to be completed for the above timeout duration. If the operation is not completed within the timeout duration, set Future of that resource in `LongRunningOperationStates` with the marshalled future data.
+* Set conditions at the end of each controller loop that describe the current state of the object and its associated Azure resources.
+
+For each Azure service, this is what the new asynchronous reconcile and delete flows will look like:
+
+##### Reconcile
+
+![Figure 1](./images/async-reconcile.png)
```

Comment:
make sure that you open up the file to view the diagrams, a lot of the proposal subject matter is in those!

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/f8276eec74b5bcea3756092f105131639c9a6874/docs/proposals/20210716-async-azure-resource-creation-deletion.md#reconcile

Reply:
Would be interesting to add a diagram for something like AzureCluster with multiple Azure resources.

Reply:
Added, PTAL

Reply:
In the diagram, "UpdateStatus" state is the only one that updates the conditions. I think "StoreLongRunningState" should probably progress to "UpdateStatus", and "UpdateStatus" probably should not set long running operation state to nil. Perhaps consider adding a state between the CreateOrUpdate state, on the success side, the long running state is set to nil, but on the context expired side, keep it the same, but progress to the "UpdateStatus" state.

Reply:
That makes sense to me